src/clr-angular/data/datagrid/providers/column-reorder.service.ts

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Injectable, ViewContainerRef, ViewRef } from '@angular/core';
+import { Observable, Subject } from 'rxjs';
+import { ColumnsService } from './columns.service';
+import { DatagridColumnChanges } from '../enums/column-changes.enum';
+
+let nbColumnsGroup = 0;
+
+@Injectable()
+export class ColumnReorderService {
+  // the common group id that will be shared across
+  // Datagrids all reorder draggable and droppables
+  private _columnsGroupId: string;
+
+  constructor(private columnsService: ColumnsService) {
+    this._columnsGroupId = 'dg-column-group-' + nbColumnsGroup++;
+  }
+
+  private reorderQueue: number[] = [];
+
+  private _reorderRequested: Subject<number[]> = new Subject<number[]>();
+
+  private _reorderCompleted: Subject<void> = new Subject<void>();
+
+  private processReorderRequest(draggedFrom: number, draggedTo: number): void {
+    // Also, the method queues all order change spec. Why do we queue each order's change spec?
+    // Because we cannot order one by one as we would lose the original order and mess up the reordering process.
+    // Before changing any column's order, we should first determine how each column's order should change first
+    // and then finally emit reorder specs array at once. The ClrDatagrid component subscribes to reorder specs event
+    // and applies the reorder specs to its columns.
+
+    this.reorderQueue = [];
+    if (draggedTo > draggedFrom) {
+      // Dragged to the right so each in-between columns should decrement their flex orders
+      for (let i = draggedFrom + 1; i <= draggedTo; i++) {
+        this.reorderQueue[i] = i - 1;
+      }
+    } else if (draggedTo < draggedFrom) {
+      // Dragged to the left so each in-between columns should decrement their flex orders
+      for (let i = draggedFrom - 1; i >= draggedTo; i--) {
+        this.reorderQueue[i] = i + 1;
+      }
+    }
+    this.reorderQueue[draggedFrom] = draggedTo;
+    // After queueing all required specs, emit what kind of reorder is requested
+    if (this.reorderQueue.length > 0) {
+      this._reorderRequested.next(this.reorderQueue);
+    }
+  }
+
+  private hasDiffWith(newOrders: number[]): boolean {
+    if (!this.orders) {
+      return true;
+    }
+    if (!newOrders) {
+      return false;
+    }
+    if (this.orders.length !== newOrders.length) {
+      return true;
+    }
+    return newOrders.filter((newOrder, index) => newOrder !== this.orders[index]).length > 0;
```

Comment:
I can't remember where I saw it but for small single typed arrays a faster solution when you want to compare them for equal order is to cast them to strings

```js
return newOrders.toString() !== this.orders.toString()
```
For a huge list, this could start to get slower, but I think is faster than looping over one array and asking to access another one and at the same time comparing them

There is no way to break at the first match - If even one is not at the correct place the result will be the same. but with the `filter` you will have to go to the end, to be absolutely sure

Another option is to use `every`

```js
return newOrders.every((newOrder, index) => newOrder !== this.orders[index]);
```

Reply:
I like your second suggestion 👍 . But you meant `.some()` instead of `.every()`, right?

Reply:
`Array.some()` is even better. I always forget about it.